pkg/kubelet/volume/volume_manager.go

```diff
@@ -0,0 +1,349 @@
+			return nil
+		}
+
+		return fmt.Errorf(
+			"timeout expired waiting for volumes to attach/mount for pod %q/%q. list of unattached/unmounted volumes=%v",
+			pod.Name,
+			pod.Namespace,
+			ummountedVolumes)
+	}
+
+	glog.Infof("All volumes are attached and mounted for pod %q", format.Pod(pod)) // TODO drop to V(3)?.
+	return nil
+}
+
+// verifyVolumesMountedFunc returns a method that returns true when all expected
+// volumes are mounted.
+func (vm *volumeManager) verifyVolumesMountedFunc(
+	podName types.UniquePodName, expectedVolumes []string) wait.ConditionFunc {
+	return func() (done bool, err error) {
+		return len(vm.getUnmountedVolumes(podName, expectedVolumes)) == 0, nil
+	}
+}
+
+// getUnmountedVolumes fetches the current list of mounted volumes from
+// the actual state of the world, and uses it to process the list of
+// expectedVolumes. It returns a list of unmounted volumes.
+func (vm *volumeManager) getUnmountedVolumes(
+	podName types.UniquePodName, expectedVolumes []string) []string {
+	mountedVolumes := sets.NewString()
+	for _, mountedVolume := range vm.actualStateOfWorld.GetMountedVolumesForPod(podName) {
+		mountedVolumes.Insert(mountedVolume.OuterVolumeSpecName)
+	}
+	return getUnmountedVolumesHelper(mountedVolumes, expectedVolumes)
+}
+
+// getUnmountedVolumesHelper adds each element of expectedVolumes that is not in
+// mountedVolumes to a list of unmountedVolumes and returns it.
+func getUnmountedVolumesHelper(
+	mountedVolumes sets.String, expectedVolumes []string) []string {
+	unmountedVolumes := []string{}
+	for _, expectedVolume := range expectedVolumes {
+		if !mountedVolumes.Has(expectedVolume) {
+			unmountedVolumes = append(unmountedVolumes, expectedVolume)
+		}
+	}
+	return unmountedVolumes
+}
+
+// getExpectedVolumes returns a list of volumes that should be mounted
+func getExpectedVolumes(pod *api.Pod) []string {
+	expectedVolumes := []string{}
+	if pod == nil {
+		return expectedVolumes
+	}
+
+	for _, podVolume := range pod.Spec.Volumes {
+		expectedVolumes = append(expectedVolumes, podVolume.Name)
+	}
+
+	return expectedVolumes
+}
+
+// getVolumesForPodHelper is a helper method for GetVolumesForPod methods.
+func (vm *volumeManager) getVolumesForPodHelper(
+	podName types.UniquePodName, pod *api.Pod) container.VolumeMap {
+	if pod != nil {
+		podName = volumehelper.GetUniquePodName(pod)
+	}
+	podVolumes := make(container.VolumeMap)
+	for _, mountedVolume := range vm.actualStateOfWorld.GetMountedVolumesForPod(podName) {
+		podVolumes[mountedVolume.OuterVolumeSpecName] =
+			container.VolumeInfo{Mounter: mountedVolume.Mounter}
+		if pod != nil {
+			err := applyPersistentVolumeAnnotations(
+				mountedVolume.VolumeGidValue, pod)
+			if err != nil {
+				glog.Errorf("applyPersistentVolumeAnnotations failed for pod %q volume %q with: %v",
+					podName,
+					mountedVolume.VolumeName,
+					err)
+			}
+		}
+	}
+	return podVolumes
+}
+
+// applyPersistentVolumeAnnotations appends a pod
```

Comment:
This is really bad.  Not your code, of course, but this is basically guaranteed to break and we need to move this into an annotation or something to avoid races / flakes.  Put an `// XXX ` comment here linking to the issue to fix this.